### PR TITLE
[Flink-1727][ml]Decision tree

### PR DESCRIPTION
**Pull request by sachingoel0101**
This implements a part of the Decision Tree Algorithm. As of now, only continuous valued fields are implemented. Also, Gini index based splitting only. Entropy to be added later.
Also adds an Online Histogram based on Ben-Haim and Yom-Tov's paper [http://www.jmlr.org/papers/volume11/ben-haim10a/ben-haim10a.pdf]

Tested on the Iris data set. [https://archive.ics.uci.edu/ml/machine-learning-databases/iris/]
Achieving an accuracy of 96.7% based on a 80:20 split of the training data. [Included in the testing suite as DecisionTreeSuite]

### CODE REVIEW

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    * *@param numClasses
+    * @return itself
+    */
+  def setClasses(numClasses: Int): DecisionTree = {
+    require(numClasses > 1, "There must be at least two classes in the data")
+    parameters.add(Classes, numClasses)
+    this
+  }
+
+  def testAccuracy(input: DataSet[LabeledVector], tree: DataSet[Tree]): Double = {
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Docstring here?

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+        tree = getRuntimeContext.getBroadcastVariable(DECISION_TREE).get(0)
+      }
+
+      override def map(labeledVector: LabeledVector): (Int, Int) = {
+        val label = tree.filter(labeledVector.vector)._2 - tree.config.labelAddition
+        if (label == labeledVector.label) (1, 1)
+        else (0, 1)
+      }
+    }
+    val result = input.map(accuracyMapper).withBroadcastSet(tree, DECISION_TREE).
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Can be changed to use the more concise mapWithBcVariable.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+import org.apache.flink.ml.common.FlinkMLTools.ModuloKeyPartitioner
+import org.apache.flink.ml.common._
+import org.apache.flink.ml.math.{CategoricalHistogram, ContinuousHistogram, OnlineHistogram, Vector}
+import org.apache.flink.ml.pipeline.{FitOperation, PredictOperation, Predictor}
+import org.apache.flink.ml.tree._
+import org.apache.flink.util.Collector
+
+import scala.collection.mutable
+
+class DecisionTree extends Predictor[DecisionTree] {
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Class docstring is required

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  /** Mapper to calculate the value of the prediction function. This is a RichMapFunction, because
+    * we broadcast the tree to all mappers.
+    */
+  class PredictionMapper[T <: Vector] extends RichMapFunction[T, LabeledVector] {
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: As with above, better to switch to mapWithBcVariable

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/tree/Tree.scala [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+import org.apache.flink.ml.math.{ContinuousHistogram, OnlineHistogram, Vector}
+
+import scala.collection.mutable
+
+/** Tree structure. This is kind of maintained in an unconventional way.
+  * We provide direct access to all nodes
+  * The obvious assumption is that child of i are 2*i and 2*i+1, while parent of i is i/2
+  *
+  */
+class Tree(
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/tree/Tree.scala]: Formatting. We try to follow the rules as presented [here](https://github.com/databricks/scala-style-guide#indent)

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/tree/Tree.scala [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+          }
+        } else {
+          finalNodes.put(x, -1)
+        }
+      }
+    )
+    evaluateNodes(finalNodes, splits, h)
+  }
+
+  private def evaluateNodes(
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/tree/Tree.scala]: Formatting here as well

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+          .cross(numberVectors)
+          .map { x => x }
+
+        instance.treeOption = Some(initialTree.iterateWithTermination(depth) {
+          tree => {
+            val currentHistograms = localHistograms(tree, blockedInputNumberElements)
+              .groupBy(0, 1, 2)
+              .reduce(
+                (a, b) => {
+                  if (a._4.isInstanceOf[ContinuousHistogram]) {
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Use pattern matching instead

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+                    (a._1, a._2, a._3, a._4.merge(b._4, maxBins + 1))
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Better to have some variable names for the sake of readability here.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+                  } else {
+                    (a._1, a._2, a._3, a._4.merge(b._4, a._4.bins))
+                  }
+                }
+              )
+            val updates = tree.map(new RichMapFunction[Tree, (Tree, Array[Int])] {
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: mapWithBcVariable

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+        })
+      }
+    }
+  }
+
+  /** Returns histograms from each input block
+    * Output format follows the form of (Node ID, Attribute Index, Label, Histogram)
+    *
+    */
+  private def localHistograms(
```
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Formatting, docstrings for parameters

**Comment by thvasilo** [comment]
Helo @sachingoel0101, I've started reviewing this PR. It would help a lot if you check out these initial comments I made and try to fix similar issues throughout the PR.

Then we can focus on the logic and design.

Also I would recommend you change the PR's name to "[FLINK-1727] [ml] Decision tree"
and add [WIP] if you consider this to be a work in progress.

Small question: I assume your are implementing the decision tree as described in the paper you linked? Could you motivate your decision to use that algorithm?

**Comment by sachingoel0101** [comment]
The fundamental idea for a scalable decision tree algorithm is to reduce the number of splits required to be checked at every node. Ideally, we'd check for every value of every attribute, which would lead to a N*D number of checks at every node [N is the number of instances and D is the dimensionality]. To bring this down, we convert the values for each attribute into a probability distribution using a histogram. After this, we can perform as many splits as we want, not depending on the actual number of training instances. Aside from this, there is nothing special in the paper which is a diversion from the standard decision tree algorithm. 

I'll update the branch soon to incorporate your comments.

**Comment by thvasilo** [comment]
On that front, do you think it's possible to split the online histogram and the decision tree code into two PRs that we can review separately?

If the online histogram code can stand on its own as a smaller feature, we can review and merge that first, and continue with the decision tree. 
What I'd like to do is break up this PR if possible, to make it easier to review and merge.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala (hunk 7 above)
Comment (sachingoel0101) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Do you mean pattern matching for checking instanceOf?

Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/DecisionTree.scala]: Yes.

**Comment by sachingoel0101** [comment]
The changes proposed by Theodore in the PR #861 have been incorporated here too. This can be reviewed now, and merging this will also close #861.

**Comment by zentol** [comment]
Closing since flink-ml is effectively frozen.
